File: core/src/main/scala/org/polars/scala/polars/api/types/DataTypes.scala
package org.polars.scala.polars.api.types

import java.time.ZoneId
import java.util.Locale
import java.util.concurrent.TimeUnit

import scala.util.Try
import scala.util.matching.Regex

trait DataType {
  def simpleName: String =
    this.getClass.getSimpleName
      .stripSuffix("$")
      .stripSuffix("Type")
      .stripSuffix("UDT")
      .toLowerCase(Locale.ROOT)
}

trait BasicDataType extends DataType

case object StringType extends BasicDataType

case object BooleanType extends BasicDataType

case object IntType extends BasicDataType

case object LongType extends BasicDataType

case object FloatType extends BasicDataType

case object DoubleType extends BasicDataType

case object DateType extends BasicDataType

case class TimeType(protected val unitStr: String) extends DataType {
  val timeUnit: Option[TimeUnit] =
    unitStr match {
      case s if s.toLowerCase(Locale.ROOT).contains("nano") => Some(TimeUnit.NANOSECONDS)
      case s if s.toLowerCase(Locale.ROOT).contains("micro") => Some(TimeUnit.MICROSECONDS)
      case s if s.toLowerCase(Locale.ROOT).contains("milli") => Some(TimeUnit.MILLISECONDS)
      case _ => None
    }

  override def simpleName: String = timeUnit match {
    case Some(TimeUnit.NANOSECONDS) => "time[ns]"
    case Some(TimeUnit.MICROSECONDS) => "time[us]"
    case Some(TimeUnit.MILLISECONDS) => "time[ms]"
    case _ => "time"
  }
}

case class DateTimeType(protected val unitStr: String, protected val tzStr: String)
    extends DataType {
  val timeUnit: Option[TimeUnit] =
    unitStr match {
      case null => None
      case s if s.toLowerCase(Locale.ROOT).contains("nano") => Some(TimeUnit.NANOSECONDS)
      case s if s.toLowerCase(Locale.ROOT).contains("micro") => Some(TimeUnit.MICROSECONDS)
      case s if s.toLowerCase(Locale.ROOT).contains("milli") => Some(TimeUnit.MILLISECONDS)
      case _ => None
    }

  val timeZone: Option[ZoneId] = Try(ZoneId.of(tzStr)).toOption

  override def simpleName: String = {
    val tu = timeUnit match {
      case Some(TimeUnit.NANOSECONDS) => "ns"
      case Some(TimeUnit.MICROSECONDS) => "us"
      case Some(TimeUnit.MILLISECONDS) => "ms"
      case _ => null
    }

    val tz = timeZone.orNull

    (tu, tz) match {
      case (null, null) => "datetime"
      case (tu, null) => s"datetime[$tu]"
      case (null, tz) => s"datetime[$tz]"
      case (tu, tz) => s"datetime[$tu, $tz]"
    }

  }
}

case class Duration(protected val unitStr: String) extends DataType {
  val timeUnit: Option[TimeUnit] =
    unitStr match {
      case s if s.toLowerCase(Locale.ROOT).contains("nano") => Some(TimeUnit.NANOSECONDS)
      case s if s.toLowerCase(Locale.ROOT).contains("micro") => Some(TimeUnit.MICROSECONDS)
      case s if s.toLowerCase(Locale.ROOT).contains("milli") => Some(TimeUnit.MILLISECONDS)
      case _ => None
    }

  override def simpleName: String = timeUnit match {
    case Some(TimeUnit.NANOSECONDS) => "duration[ns]"
    case Some(TimeUnit.MICROSECONDS) => "duration[us]"
    case Some(TimeUnit.MILLISECONDS) => "duration[ms]"
    case _ => "duration"
  }
}

case class ListType(tpe: DataType) extends DataType {
  override def simpleName: String = "list"

  /** Borrowed from Apache Spark source to represent [[ListType]] as a tree string. */
  private[polars] def buildFormattedString(prefix: String, buffer: StringBuffer): Unit = {
    buffer.append(s"$prefix-- element: ${tpe.simpleName}\n")
    DataType.buildFormattedString(tpe, s"$prefix    |", buffer)
  }

}

case class StructType(fields: Seq[Field]) extends DataType {
  override def simpleName: String = "struct"

  /** Borrowed from Apache Spark source to represent [[StructType]] as a tree string. */
  private[polars] def buildFormattedString(prefix: String, buffer: StringBuffer): Unit =
    fields.foreach(field => field.buildFormattedString(prefix, buffer))
}

object DataType {

  private[polars] final val StringRegex: Regex = """^(?i)Utf8|LargeUtf8|String$""".r
  private[polars] final val BooleanRegex: Regex = """^(?i)Boolean$""".r
  private[polars] final val IntRegex: Regex = """^(?i)Int8|Int16|Int32|UInt8|UInt16|UInt32$""".r
  private[polars] final val LongRegex: Regex = """^(?i)Int64|UInt64$""".r
  private[polars] final val FloatRegex: Regex = """^(?i)Float32$""".r
  private[polars] final val DoubleRegex: Regex = """^(?i)Float64$""".r
  private[polars] final val DateRegex: Regex = """^(?i)Date|Date32|Date64$""".r

  def fromBasicType(typeStr: String): DataType = typeStr match {
    case StringRegex() => StringType
    case BooleanRegex() => BooleanType
    case IntRegex() => IntType
    case LongRegex() => LongType
    case FloatRegex() => FloatType
    case DoubleRegex() => DoubleType
    case DateRegex() => DateType
    case typeStr =>
      throw new IllegalArgumentException(s"Unknown basic type `$typeStr` is not supported.")
  }

  /** Borrowed from Apache Spark source to represent [[DataType]] as a tree string. */
  private[polars] def buildFormattedString(
      dataType: DataType,
      prefix: String,
      buffer: StringBuffer
  ): Unit =
    dataType match {
      case array: ListType => array.buildFormattedString(prefix, buffer)
      case struct: StructType => struct.buildFormattedString(prefix, buffer)
      case _ =>
    }

}

File: core/src/main/scala/com/github/chitralverma/polars/api/types/DataTypes.java
package com.github.chitralverma.polars.api.types;

/**
 * Spark-style Java-friendly facade for all Polars DataTypes.
 * This class exposes public static final fields and static factory methods
 * so Java and Scala users can refer to types cleanly (e.g. {@code DataTypes.Int32}
 * instead of leaking Scala objects).
 */
public final class DataTypes {

  private DataTypes() {}

  // Basic singletons (Polars-faithful names)
  public static final DataType String = StringType$.MODULE$;
  public static final DataType Boolean = BooleanType$.MODULE$;
  public static final DataType Int8 = Int8Type$.MODULE$;
  public static final DataType Int16 = Int16Type$.MODULE$;
  public static final DataType Int32 = Int32Type$.MODULE$;
  public static final DataType Int64 = Int64Type$.MODULE$;
  public static final DataType UInt8 = UInt8Type$.MODULE$;
  public static final DataType UInt16 = UInt16Type$.MODULE$;
  public static final DataType UInt32 = UInt32Type$.MODULE$;
  public static final DataType UInt64 = UInt64Type$.MODULE$;
  public static final DataType Float32 = Float32Type$.MODULE$;
  public static final DataType Float64 = Float64Type$.MODULE$;
  public static final DataType Date = DateType$.MODULE$;
  public static final DataType Time = TimeType$.MODULE$;
  public static final DataType Null = NullType$.MODULE$;

  // Default DecimalType singleton (precision=None, scale=0)
  public static final DataType Decimal = new DecimalType(scala.Option.apply(null), 0);

  // Static Factory Methods for parametric types (S2 / Q2)

  /**
   * Create a Decimal DataType with the given scale.
   *
   * @param scale number of digits after the decimal point
   * @return Decimal DataType
   */
  public static DataType decimal(int scale) {
    return new DecimalType(scala.Option.apply(null), scale);
  }

  /**
   * Create a Decimal DataType with the given precision and scale.
   *
   * @param precision total number of digits
   * @param scale number of digits after the decimal point
   * @return Decimal DataType
   */
  public static DataType decimal(int precision, int scale) {
    return new DecimalType(scala.Option.apply(precision), scale);
  }

  /**
   * Create a Time DataType with the specified time unit.
   *
   * @param timeUnit time unit (e.g. "Microseconds", "Nanoseconds")
   * @return Time DataType
   */
  public static DataType time(String timeUnit) {
    return new TimeType(timeUnit);
  }

  /**
   * Create a Datetime DataType with the specified time unit and timezone.
   *
   * @param timeUnit time unit (e.g. "Microseconds")
   * @param timeZone timezone identifier (e.g. "UTC", "Asia/Tokyo") or null
   * @return Datetime DataType
   */
  public static DataType datetime(String timeUnit, String timeZone) {
    return new DateTimeType(timeUnit, timeZone);
  }

  /**
   * Create a Duration DataType with the specified time unit.
   *
   * @param timeUnit time unit (e.g. "Microseconds")
   * @return Duration DataType
   */
  public static DataType duration(String timeUnit) {
    return new Duration(timeUnit);
  }

  /**
   * Create a List DataType with the specified element type.
   *
   * @param elementType the DataType of the list's elements
   * @return List DataType
   */
  public static DataType list(DataType elementType) {
    return new ListType(elementType);
  }

  /**
   * Create a Struct DataType consisting of the specified fields.
   *
   * @param fields fields defining the struct's schema
   * @return Struct DataType
   */
  public static DataType struct(Field[] fields) {
    return new StructType(fields);
  }
}
